smack-extensions/src/main/java/org/jivesoftware/smackx/avatar/UserAvatarManager.java

```diff
@@ -0,0 +1,318 @@
+     *
+     * @param connection
+     * @return the instance of UserAvatarManager
+     */
+    public static synchronized UserAvatarManager getInstanceFor(XMPPConnection connection) {
+        UserAvatarManager userAvatarManager = INSTANCES.get(connection);
+
+        if (userAvatarManager == null) {
+            userAvatarManager = new UserAvatarManager(connection);
+            INSTANCES.put(connection, userAvatarManager);
+        }
+
+        return userAvatarManager;
+    }
+
+    private UserAvatarManager(XMPPConnection connection) {
+        super(connection);
+        this.pepManager = PepManager.getInstanceFor(connection);
+    }
+
+    /**
+     * Returns true if User Avatar is supported by the server.
+     *
+     * @return true if User Avatar is supported by the server.
+     * @throws NoResponseException
+     * @throws XMPPErrorException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     */
+    public boolean isSupportedByServer()
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        return findNodeItem(DATA_NAMESPACE) && findNodeItem(METADATA_NAMESPACE);
+    }
+
+    private boolean findNodeItem(String nodeName)
+            throws NoResponseException, XMPPErrorException, NotConnectedException, InterruptedException {
+        ServiceDiscoveryManager serviceDiscoveryManager = ServiceDiscoveryManager.getInstanceFor(connection());
+        List<Item> items = serviceDiscoveryManager.discoverItems(connection().getUser().asBareJid(), nodeName)
+                .getItems();
+        return items != null && items.size() > 0;
+    }
+
+    /**
+     * Get the data node.
+     *
+     * @return the data node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getDataNode()
+            throws NoResponseException, NotConnectedException, InterruptedException, XMPPErrorException, PubSubException.NotALeafNodeException {
+        return pepManager.getPepPubSubManager().getOrCreateLeafNode(DATA_NAMESPACE);
+    }
+
+    /**
+     * Get the metadata node.
+     *
+     * @return the metadata node
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws InterruptedException
+     * @throws XMPPErrorException
+     */
+    public LeafNode getMetadataNode()
+            throws NoResponseException, NotConnectedException, InterruptedException, XMPPErrorException, PubSubException.NotALeafNodeException {
+        return pepManager.getPepPubSubManager().getOrCreateLeafNode(METADATA_NAMESPACE);
+    }
+
+    /**
+     * Publish avatar data.
+     *
+     * @param data
+     * @param itemId
+     * @throws NoResponseException
+     * @throws NotConnectedException
+     * @throws XMPPErrorException
+     * @throws InterruptedException
+     */
+    public void publishAvatarData(byte[] data, String itemId)
```

Comment:
We probably don't want to let the user deal with itemIDs.
So, private it is?

Reply:
Make it so, we can later make it public if it is really needed. But a good API has as less as possible public methods.